src/interpreter/mod.rs

```diff
@@ -173,11 +173,27 @@ impl<R: RealNumberInternalTrait> std::ops::Div<Number<R>> for Number<R> {
 
 pub type ArgVec<R, E> = SmallVec<[Value<R, E>; 4]>;
 
-#[derive(Clone)]
+#[derive(Clone, PartialEq)]
+pub enum BuildinProcedurePointer<R: RealNumberInternalTrait, E: IEnvironment<R>> {
+    Pure(fn(ArgVec<R, E>) -> Result<Value<R, E>>),
+    Impure(
+        fn(ArgVec<R, E>, Rc<RefCell<E>>) -> Result<Value<R, E>>,
+        Rc<RefCell<E>>,
+    ),
+}
+impl<R: RealNumberInternalTrait, E: IEnvironment<R>> BuildinProcedurePointer<R, E> {
+    pub fn eval(&self, args: ArgVec<R, E>) -> Result<Value<R, E>> {
```

Comment:
BTW, `eval` is for a expression . If you eval a procedure it means you get the procedure itself. Use `apply` or `call` instead